larpandoracontent/LArControlFlow/TrackDirectionTool.cc

```diff
@@ -0,0 +1,1154 @@
+	{
+	  throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+	}
+
+      const Cluster *const pCluster(*(bestClusterList.begin())); 
+      return pCluster;
+
+    }
+    else{
+      throw StatusCodeException(STATUS_CODE_FAILURE);
+      return 0;
+    }
+}
+
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::SetEndpoints(DirectionFitObject &fitResult, const Cluster *const pCluster)
+{
+    CartesianVector lowZVector(0.f, 0.f, 0.f), highZVector(0.f, 0.f, 0.f);
+    LArClusterHelper::GetExtremalCoordinates(pCluster, lowZVector, highZVector);
+
+    fitResult.SetBeginpoint(lowZVector);
+    fitResult.SetEndpoint(highZVector);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::SetEndpoints(DirectionFitObject &fitResult, const LArTrackStateVector &trackStateVector)
+{
+    TrackState firstTrackState(*(trackStateVector.begin())), lastTrackState(trackStateVector.back());
+    const pandora::CartesianVector initialPosition(firstTrackState.GetPosition());
+    const pandora::CartesianVector endPosition(lastTrackState.GetPosition());
+
+    const pandora::CartesianVector lowZVector(initialPosition.GetZ() < endPosition.GetZ() ? initialPosition : endPosition);
+    const pandora::CartesianVector highZVector(initialPosition.GetZ() > endPosition.GetZ() ? initialPosition : endPosition);
+
+    fitResult.SetBeginpoint(lowZVector);
+    fitResult.SetEndpoint(highZVector);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::FillHitChargeVector(const Cluster *const pCluster, HitChargeVector &hitChargeVector)
+{
+    OrderedCaloHitList orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+    CaloHitList caloHitList;
+    orderedCaloHitList.FillCaloHitList(caloHitList);
+
+    const TwoDSlidingFitResult &slidingFit(this->GetCachedSlidingFit(pCluster));
+
+    for (CaloHitList::const_iterator hitIter = caloHitList.begin(), hitIterEnd = caloHitList.end(); hitIter != hitIterEnd; ++hitIter)
+    {
+        const CaloHit *const pCaloHit(*hitIter);
+        const CartesianVector caloHitPosition(pCaloHit->GetPositionVector());
+        float hitWidth(pCaloHit->GetCellSize1());
+
+        float caloHitEnergy(pCaloHit->GetInputEnergy());
+        caloHitEnergy *= 273.5; //ADC to electron
+        caloHitEnergy *= 23.6/1000000; //ionisation energy per electron in MeV
+        caloHitEnergy /= 0.62;
+
+        float rL(0.f), rT(0.f);
+        slidingFit.GetLocalPosition(caloHitPosition, rL, rT);
+        if (rL == 0.)
+            continue;
+
+        float calibratedUncertainty(std::sqrt((0.00419133 * (caloHitEnergy/hitWidth) * (caloHitEnergy/hitWidth)) + (0.00967141 * (caloHitEnergy/hitWidth)))); //70%
```

Comment:
But they're fixed as calibration parameters?

Reply:
It's always possible that calibration parameters could be altered in future with any recalibration, and different experiments could have quite different calibrations. It's fine to set these values as the default for any configurable variables to avoid the need to specify them in XML for MicroBooNE (similarly for the other parameters discussed elsewhere), and have future updates/different experiments overriding them as needed. The primary goal here is to ensure that in any future work in this area, it's clear what the numbers 0.00419 and 0.00967 are and where they come from, in case any other use case needs to use different values.

Reply:
Right okay, done